File: app/streamlit_funnels/demands/demandes_repo.py
"""Supabase access for agence_demandes carousel cards."""

from __future__ import annotations

from functools import lru_cache
from typing import Any

from config import require_supabase
from supabase import Client, create_client


@lru_cache(maxsize=1)
def get_client() -> Client:
    url, key = require_supabase()
    return create_client(url, key)


def list_all_cards() -> list[dict[str, Any]]:
    resp = (
        get_client()
        .table("agence_demandes")
        .select("*")
        .order("sort_order")
        .execute()
    )
    return resp.data or []


def list_demandes() -> list[dict[str, Any]]:
    return [row for row in list_all_cards() if row.get("record_type") == "demande"]


def list_teasers() -> list[dict[str, Any]]:
    return [row for row in list_all_cards() if row.get("record_type") == "teaser"]


def get_card(external_id: str) -> dict[str, Any] | None:
    resp = (
        get_client()
        .table("agence_demandes")
        .select("*")
        .eq("external_id", external_id)
        .maybe_single()
        .execute()
    )
    return resp.data if resp else None


def update_demande(external_id: str, fields: dict[str, Any]) -> None:
    allowed = {
        "niche",
        "secteur",
        "prestation",
        "budget",
        "taille",
        "zone",
        "disponibilite",
        "origine",
        "status",
        "available_from",
        "available_until",
    }
    payload = {key: value for key, value in fields.items() if key in allowed}
    if not payload:
        return
    get_client().table("agence_demandes").update(payload).eq("external_id", external_id).execute()


def update_teaser(external_id: str, fields: dict[str, Any]) -> None:
    allowed = {"secteur", "titre", "description", "note"}
    payload = {key: value for key, value in fields.items() if key in allowed}
    if not payload:
        return
    get_client().table("agence_demandes").update(payload).eq("external_id", external_id).execute()

File: lib/site/comptable/demandes-repo.ts
import type {
  DemandeContrat,
  DemandeNiche,
  DemandeStatus,
  DemandeTeaser,
} from "@/lib/site/demandes-data";
import { withMarketingFetchFallback } from "@/lib/site/marketing/resilient-fetch";

interface ComptableDemandeRow {
  external_id: string;
  record_type: "demande" | "teaser";
  niche: DemandeNiche;
  secteur: string;
  prestation: string | null;
  budget: string | null;
  taille: string | null;
  zone: string | null;
  disponibilite: string | null;
  origine: string | null;
  duree_souhaitee: string | null;
  horizon_resultat: string | null;
  historique_agences: string | null;
  status: DemandeStatus | null;
  available_from: string | null;
  available_until: string | null;
  titre: string | null;
  description: string | null;
  note: string | null;
  sort_order: number;
}

export function mapComptableDemandeRow(row: ComptableDemandeRow): DemandeContrat {
  return {
    id: row.external_id,
    niche: row.niche,
    secteur: row.secteur,
    availableFrom: row.available_from ?? "",
    availableUntil: row.available_until ?? "",
    prestation: row.prestation ?? "",
    budget: row.budget ?? "",
    taille: row.taille ?? "",
    dureeSouhaitee: row.duree_souhaitee ?? "",
    horizonResultat: row.horizon_resultat ?? "",
    historiqueAgences: row.historique_agences ?? "",
    zone: row.zone ?? "",
    disponibilite: row.disponibilite ?? "",
    origine: row.origine ?? "",
    status: row.status ?? "available",
    masked: true,
  };
}

function mapTeaserRow(row: ComptableDemandeRow): DemandeTeaser {
  return {
    id: row.external_id,
    niche: "a-venir",
    secteur: row.secteur,
    titre: row.titre ?? "",
    description: row.description ?? "",
    note: row.note ?? "",
  };
}

async function loadComptableDemandesForCarousel(): Promise<DemandeContrat[]> {
  return [];
}

export async function fetchComptableDemandesForCarousel(): Promise<DemandeContrat[]> {
  return withMarketingFetchFallback(
    "comptable carousel demandes",
    loadComptableDemandesForCarousel,
    [],
  );
}

async function loadComptableDemandeTeaser(): Promise<DemandeTeaser | null> {
  return null;
}

export async function fetchComptableDemandeTeaser(): Promise<DemandeTeaser | null> {
  return withMarketingFetchFallback(
    "comptable demande teaser",
    loadComptableDemandeTeaser,
    null,
  );
}
